### PR TITLE
Receiving Refactor: handleDataMessage onwards

### PR DESCRIPTION
**Pull request by msgmaxim**

### CODE REVIEW

File: ts/receiver/types.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    OUTGOING = "outgoing",
+    FRIEND_REQUEST = "friend-request"
+}
+
+export enum EndSessionType {
+    DONE = 'done',
+    ONGOING = 'ongoing',
+}
+
+interface MessageAttributes {
```
Comment (Mikunj) [inline ts/receiver/types.ts:13]: this should probably be put in `messages.d.ts` next to `messages.js` file in models

File: ts/receiver/types.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    isGroupUpdate: () => boolean;
+    isExpirationTimerUpdate: () => boolean;
+    isFriendRequest: () => boolean;
+    getNotificationText: () => string;
+    isEndSession: () => boolean;
+    markRead: () => void;
+    merge: (other: MessageModel) => void;
+}
+
+interface ConversationAttributes {
```
Comment (Mikunj) [inline ts/receiver/types.ts:59]: same as above but in `conversation.d.ts`

File: ts/receiver/receiver.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+    const type = isGroupMessage ? ConversationType.GROUP : ConversationType.PRIVATE;
+
+    // MAXIM: So id is actually conversationId
+    const id = isIncoming ? source : destination;
+
+    const {
+        PROFILE_KEY_UPDATE,
+        SESSION_REQUEST,
+        SESSION_RESTORE,
+    } = window.textsecure.protobuf.DataMessage.Flags;
```
Comment (Mikunj) [inline ts/receiver/receiver.ts:335]: `window.textsecure.protobuf` can be replaced with `SignalService` directly

File: ts/receiver/types.ts (hunk 1 above)
Comment (msgmaxim) [reply ts/receiver/types.ts]: done

File: ts/receiver/receiver.ts (hunk 3 above)
Comment (msgmaxim) [reply ts/receiver/receiver.ts]: done

File: ts/receiver/types.ts (hunk 2 above)
Comment (msgmaxim) [reply ts/receiver/types.ts]: done

File: ts/receiver/attachments.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  }
+
+  // The attachment id is actually just the absolute url of the attachment
+  let data = await webAPI.getAttachment(attachment.url);
+  if (!attachment.isRaw) {
+    const { key, digest, size } = attachment;
+
+    data = await window.textsecure.crypto.decryptAttachment(
+      data,
+      window.Signal.Crypto.base64ToArrayBuffer(key),
```
Comment (Mikunj) [inline ts/receiver/attachments.ts:20]: you can do
```
Buffer.from(key, "base64").toArrayBuffer()
```

File: ts/receiver/conversation.d.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+interface ConversationAttributes {
```
Comment (Mikunj) [inline ts/receiver/conversation.d.ts:1]: for this i meant put it in
`js/models/conversation.d.ts`

File: ts/receiver/conversation.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import { MessageModel } from './message';
+
+interface ConversationAttributes {
```
Comment (Mikunj) [inline ts/receiver/conversation.ts:3]: hmm why is this file needed?

File: ts/receiver/message.d.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
+enum MessageModelType {
```
Comment (Mikunj) [inline ts/receiver/message.d.ts:1]: same here, this should be in
`js/models/message.d.ts`

File: ts/receiver/message.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
+enum MessageModelType {
```
Comment (Mikunj) [inline ts/receiver/message.ts:1]: again why do we need this file?

File: ts/receiver/multidevice.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  return !found;
+}
+
+async function clearAppAndRestart() {
+  // remove our device mapping annotations from file server
+  await window.lokiFileServerAPI.clearOurDeviceMappingAnnotations();
+  // Delete the account and restart
+  try {
+    await window.Signal.Logs.deleteAll();
+    await window.Signal.Data.removeAll();
```
Comment (Mikunj) [inline ts/receiver/multidevice.ts:42]: these calls to Data can be imported directly.
```
import { removeAll } from 'js/modules/data.js';
```
just keep in mind doing:
```
import * as Data from 'js/modules.data'
```
will make it so stubbing those data functions in tests won't be possible. So if this code isn't going to be tested yet we might want to use the latter for now.

File: ts/receiver/receiver.ts [hunk 10]
```diff
@@ -0,0 +1,824 @@
+// TODO: fix libloki and textsecure not being available here yet
```
Comment (Mikunj) [inline ts/receiver/receiver.ts:1]: actually a lot of these should be available if you put them in `window/index.ts`.
Then you can just import like normal.

```
import { libloki, textsecure } from '../window'
```
We're going to be removing `global.d.ts` so using `window.<variable>` won't be allowed.

**Comment by Mikunj** [comment]
An alternative we can do is merge this in and in a separate PR you can refine those specific ts changes (importing window, etc..). This will make it so we won't have many conflicts when we work on the receiver for other refactors

File: ts/receiver/receiver.ts [hunk 11]
```diff
@@ -0,0 +1,824 @@
...
+
+    // Will do the save for us
+    await sender.setProfileKey(profileKey);
+  }
+}
+
+enum ConversationType {
+  GROUP = 'group',
+  PRIVATE = 'private',
+}
```
Comment (Bilb) [inline ts/receiver/receiver.ts:176]: same than below

File: ts/receiver/types.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { SignalService } from '../protobuf';
+
+export interface Quote {
+  id: any;
+  author: any;
+  attachments: Array<any>;
+  text: string;
+  referencedMessageNotFound: boolean;
+}
```
Comment (Bilb) [inline ts/receiver/types.ts:9]: in ChatMessage.ts  I have a declaration of 
```
export interface Quote {
  id?: number;
  author?: string;
  text?: string;
  attachments?: Array<QuotedAttachment>;
```
with the type description of a QuotedAttachment too. Should those two interface be merged in a single one?

File: ts/receiver/conversation.ts (hunk 6 above)
Comment (msgmaxim) [reply ts/receiver/conversation.ts]: removed

File: ts/receiver/message.ts (hunk 8 above)
Comment (msgmaxim) [reply ts/receiver/message.ts]: removed

File: ts/receiver/message.d.ts (hunk 7 above)
Comment (msgmaxim) [reply ts/receiver/message.d.ts]: moved

File: ts/receiver/conversation.d.ts (hunk 5 above)
Comment (msgmaxim) [reply ts/receiver/conversation.d.ts]: moved

File: ts/receiver/receiver.ts [hunk 13]
```diff
@@ -0,0 +1,824 @@
...
+      error && error.stack ? error.stack : error
+    );
+  });
+}
+
+interface MessageId {
+  source: any;
+  sourceDevice: any;
+  timestamp: any;
+}
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply ts/receiver/receiver.ts:200]: These are likely temporary (and not used anywhere else), so don't think we should move them.

File: ts/receiver/receiver.ts (hunk 11 above)
Comment (msgmaxim) [reply ts/receiver/receiver.ts:176]: (as above)

File: ts/receiver/types.ts (hunk 12 above)
Comment (msgmaxim) [reply ts/receiver/types.ts:9]: Quote has extra fileds on top of what is in the protobuf. Also I think we should use a tighter type for Quote (and others) inside business logic and only deal with protobuf types on the network boundary.
